compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp

```diff
@@ -377,6 +385,63 @@ class FuseTransposeWithGenericConsumer
   }
 };
 
+// Propagates a transpose through the input of a unary elementwise operation.
+class PropagateTransposeThroughUnaryElementwiseInput
+    : public OpRewritePattern<linalg::GenericOp> {
+public:
+  using OpRewritePattern<linalg::GenericOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::GenericOp genericOp,
+                                PatternRewriter &rewriter) const override {
+    if (!IREE::Flow::isNonNullAndOutsideDispatch(genericOp)) {
+      return failure();
+    }
+
+    if (genericOp.getNumDpsInputs() != 1 || genericOp.getNumDpsInputs() != 1 ||
+        !linalg::isElementwise(genericOp)) {
+      return rewriter.notifyMatchFailure(genericOp, "not unary elementwise");
+    }
+
+    OpOperand *input = genericOp.getDpsInputOperand(0);
+    OpOperand *init = genericOp.getDpsInitOperand(0);
+
+    // Skip transposes and broadcasts. Transposes make more sense to fuse
+    // rather than propagate through, and broadcasts are cheaper to transpose
+    // before broadcasting.
+    if (genericOp.getMatchingIndexingMap(input) !=
+        genericOp.getMatchingIndexingMap(init)) {
+      return failure();
+    }
+
+    linalg::TransposeOp transposeOp =
+        input->get().getDefiningOp<linalg::TransposeOp>();
+    if (!transposeOp) {
+      return rewriter.notifyMatchFailure(genericOp, "no transpose operand");
+    }
+
+    ArrayRef<int64_t> perm = transposeOp.getPermutation();
+    auto invPerm = invertPermutationVector(perm);
+
+    // Create a new empty init for the transposed generic.
+    Value newInit = createTransposeInit(rewriter, init->get(), invPerm);
+
+    // We do not need to update indexing maps because this is a unary
+    // elementwise op where the input and output maps are the same. Just
+    // replace the operands with transposed variants.
+    auto newGenericOp = rewriter.create<linalg::GenericOp>(
+        genericOp.getLoc(), newInit.getType(), transposeOp.getInput(), newInit,
+        genericOp.getIndexingMapsArray(), genericOp.getIteratorTypesArray(),
+        /*bodyBuild=*/nullptr, linalg::getPrunedAttributeList(genericOp));
+    rewriter.cloneRegionBefore(genericOp.getRegion(), newGenericOp.getRegion(),
+                               newGenericOp.getRegion().begin());
```

Comment:
Can we use `getInputsMutable` and `getOutputsMutable` here?

```
genericOp.getInputsMutable().assign(transposeOp.getInput());
genericOp.getOutputMutable().assign(newInit);
```

Reply:
I don't think so because I need to change the result type of the generic too. Is there a way to update the result type of an operation in place?

Reply:
I see, sorry for the bad suggestion. Are we able to use `clone` method here? It passes result types and new operands. The region is fine because they are identical right?

https://github.com/llvm/llvm-project/blob/1e8d3c357e1fd79bf3a641767d95147a8c0f54bd/mlir/include/mlir/Dialect/Utils/StructuredOpsUtils.h#L139-L143

Reply:
No worries. Clone looks better, thanks!